marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/entities/entity-search-lib.sjs

```diff
@@ -535,11 +537,94 @@ function isHubEntityInstance(docUri) {
+      history.push(historyObject);
+    });
+  } else {
+    const metadata = getRecordMetadata(docUri);
+    if(Object.keys(metadata).length) {
+      history.push(getRecordMetadata(docUri));
+    }
+  }
+  return history;
+}
+
+function findFlowAndStepNameFromProvenanceRecords(provenanceRecord, flowNames, stepNames, stepDefinitionNames) {
+  const flowAndStepNames = {};
+  const associatedWith = provenanceRecord.associatedWith.toObject();
+  const commonFlowNames = associatedWith.filter(artifactName => flowNames.includes(artifactName.toString()));
+
+  commonFlowNames.forEach(artifactName => {
+    if(provenanceRecord["provID"].toString().includes(artifactName)) {
+      flowAndStepNames["flowName"] = artifactName;
+      for (let currentIndex=0; currentIndex<associatedWith.length-1; currentIndex++) {
+        if(associatedWith[currentIndex] === artifactName) {
+          let artifactTobeRemoved =  associatedWith[currentIndex];
+          associatedWith[currentIndex] = associatedWith[associatedWith.length-1];
+          associatedWith[associatedWith.length-1] = artifactTobeRemoved;
+          break;
+        }
+      }
+      associatedWith.pop();
+    }
+  });
+
+  const commonStepNames = associatedWith.filter(artifactName => stepNames.includes(artifactName));
+  flowAndStepNames["stepName"] = commonStepNames.length ? commonStepNames[0] : undefined;
+  if(commonStepNames.length == 0) {
+    const commonStepDefinitionNames = associatedWith.filter(artifactName => stepDefinitionNames.includes(artifactName));
+    flowAndStepNames["stepName"] = commonStepDefinitionNames.length ? commonStepDefinitionNames[0] : undefined;
+  }
+
+  return flowAndStepNames;
+}
+
+function getRecordMetadata(docUri) {
+  const metadata = xdmp.documentGetMetadata(docUri);
+  const currentObject = {};
+  if(metadata) {
+    // datahubCreatedOn field captures the document/record last updated timestamp. It is not the document/record creation timestamp
+    currentObject.updatedTime = metadata.datahubCreatedOn ? metadata.datahubCreatedOn : undefined;
+    currentObject.flow = metadata.datahubCreatedInFlow ? metadata.datahubCreatedInFlow : undefined;
+    currentObject.step = metadata.datahubCreatedByStep ? metadata.datahubCreatedByStep : undefined;
+    currentObject.user = metadata.datahubCreatedBy ? metadata.datahubCreatedBy : undefined;
+  }
+  return currentObject;
+}
+
+function getArtifactNamesFromUris(collection, expectedPrefix, expectedSuffix) {
+  const artifactNames = [];
+  cts.uris(null, null, cts.collectionQuery(collection)).toArray().map(uri => {
+    uri = uri.toString();
+    if (uri.startsWith(expectedPrefix) && uri.endsWith(expectedSuffix)) {
+      artifactNames.push(uri.replace(expectedPrefix,'').replace(new RegExp(expectedSuffix + '$'), '').split("/").pop());
+    }
+  });
+  return artifactNames;
+}
+
 module.exports = {
   addDocumentMetadataToSearchResults,
   addPropertiesToSearchResponse,
   buildPropertyMetadata,
+  getArtifactNamesFromUris,
```

Comment:
I didn't see anywhere that this needed to be exported, I think it can be private to this library.